src/libstd/io/buffered.rs

```diff
@@ -0,0 +1,637 @@
+    pub fn get_ref(&self) -> &W { self.inner.as_ref().unwrap() }
+
+    /// Gets a mutable reference to the underlying write.
+    ///
+    /// # Warning
+    ///
+    /// It is inadvisable to directly read from the underlying writer.
+    pub fn get_mut(&mut self) -> &mut W { self.inner.as_mut().unwrap() }
+
+    /// Unwraps this `BufWriter`, returning the underlying writer.
+    ///
+    /// The buffer is flushed before returning the writer.
+    pub fn into_inner(mut self) -> W {
+        // FIXME(#12628): is panicking the right thing to do if flushing panicks?
+        self.flush_buf().unwrap();
+        self.inner.take().unwrap()
+    }
+}
+
+impl<W: Write> Write for BufWriter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.buf.len() + buf.len() > self.buf.capacity() {
+            try!(self.flush_buf());
+        }
+        if self.buf.len() == 0 && buf.len() >= self.buf.capacity() {
+            self.inner.as_mut().unwrap().write(buf)
+        } else {
+            let amt = cmp::min(buf.len(), self.buf.capacity());
+            Write::write(&mut self.buf, &buf[..amt])
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.flush_buf().and_then(|_| self.get_mut().flush())
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<W> fmt::Debug for BufWriter<W> where W: fmt::Debug {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "BufWriter {{ writer: {:?}, buffer: {}/{} }}",
+               self.inner.as_ref().unwrap(), self.buf.len(), self.buf.capacity())
+    }
+}
+
+#[unsafe_destructor]
+impl<W: Write> Drop for BufWriter<W> {
+    fn drop(&mut self) {
+        if self.inner.is_some() {
+            // dtors should not panic, so we ignore a panicked flush
+            let _ = self.flush_buf();
+        }
+    }
+}
+
+/// Wraps a Writer and buffers output to it, flushing whenever a newline
+/// (`0x0a`, `'\n'`) is detected.
+///
+/// This writer will be flushed when it is dropped.
+pub struct LineWriter<W> {
+    inner: BufWriter<W>,
+}
+
+impl<W: Write> LineWriter<W> {
+    /// Creates a new `LineWriter`
+    pub fn new(inner: W) -> LineWriter<W> {
+        // Lines typically aren't that long, don't use a giant buffer
+        LineWriter { inner: BufWriter::with_capacity(1024, inner) }
+    }
+
+    /// Gets a reference to the underlying writer.
+    ///
+    /// This type does not expose the ability to get a mutable reference to the
+    /// underlying reader because that could possibly corrupt the buffer.
+    pub fn get_ref<'a>(&'a self) -> &'a W { self.inner.get_ref() }
+
+    /// Unwraps this `LineWriter`, returning the underlying writer.
+    ///
+    /// The internal buffer is flushed before returning the writer.
+    pub fn into_inner(self) -> W { self.inner.into_inner() }
+}
+
+impl<W: Write> Write for LineWriter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match buf.rposition_elem(&b'\n') {
```

Comment:
I agree that this is definitely an area where we can have future optimizations. For now I think we may want to track this in an issue because we're not exposing `memchr` in the standard library (or `memrchr`). I tried to move instances of `.iter().position(..)` to `position_elem` or `rposition_elem` so they're easy to flag and migrate once we have them implemented.

One possibility would be to expand the [`slice::bytes`](http://doc.rust-lang.org/std/slice/bytes/) module with some functions like this as a place to house the optimized versions.